common/kernel/timing.cc

```diff
@@ -742,545 +782,75 @@ std::vector<CellPortKey> TimingAnalyser::get_failing_eps(domain_id_t domain_pair
-                        continue;
-                    DelayQuad comb_delay;
-                    bool is_path = ctx->getCellDelay(usr.cell, usr.port, port.first, comb_delay);
-                    if (!is_path)
-                        continue;
-                    // Decrement the fanin count, and only add to topological order if all its fanins have already
-                    // been visited
-                    auto it = port_fanin.find(&port.second);
-                    if (it == port_fanin.end())
-                        log_error("Timing counted negative fanin count for port %s.%s (net %s), please report this "
-                                  "error.\n",
-                                  ctx->nameOf(usr.cell), ctx->nameOf(port.first), ctx->nameOf(port.second.net));
-                    if (--it->second == 0) {
-                        topological_order.emplace_back(port.second.net);
-                        queue.emplace_back(port.second.net);
-                        port_fanin.erase(it);
-                    }
-                }
-            }
+    if (!launch.is_async() && ctx->nets.at(launch.clock)->clkconstr) {
+        if (launch.edge == capture.edge) {
+            report.period = ctx->nets.at(launch.clock)->clkconstr->period.minDelay();
+        } else if (capture.edge == RISING_EDGE) {
+            report.period = ctx->nets.at(launch.clock)->clkconstr->low.minDelay();
+        } else if (capture.edge == FALLING_EDGE) {
+            report.period = ctx->nets.at(launch.clock)->clkconstr->high.minDelay();
         }
+    }
 
-        // Sanity check to ensure that all ports where fanins were recorded were indeed visited
-        if (!port_fanin.empty() && !bool_or_default(ctx->settings, ctx->id("timing/ignoreLoops"), false)) {
-            for (auto fanin : port_fanin) {
-                NetInfo *net = fanin.first->net;
-                if (net != nullptr) {
-                    log_info("   remaining fanin includes %s (net %s)\n", fanin.first->name.c_str(ctx),
-                             net->name.c_str(ctx));
-                    if (net->driver.cell != nullptr)
-                        log_info("        driver = %s.%s\n", net->driver.cell->name.c_str(ctx),
-                                 net->driver.port.c_str(ctx));
-                    for (auto net_user : net->users)
-                        log_info("        user: %s.%s\n", net_user.cell->name.c_str(ctx), net_user.port.c_str(ctx));
-                } else {
-                    log_info("   remaining fanin includes %s (no net)\n", fanin.first->name.c_str(ctx));
-                }
-            }
-            if (ctx->force)
-                log_warning("timing analysis failed due to presence of combinatorial loops, incomplete specification "
-                            "of timing ports, etc.\n");
-            else
-                log_error("timing analysis failed due to presence of combinatorial loops, incomplete specification of "
```

Comment:
If I haven't missed anything in your new code, I think you will need to re-add the loop error (albeit ideally with better reporting of the loop now we have toposort - see the current loop reporting code in the new timing analyser) and override with ignoreLoops.

Reply:
Yes, you are right. I half expected the logs in the current analyser to be enough. But not having warnings/errors is a big regressions we shouldn't make.

Reply:
This is added back now. I have removed the `verbose_mode` guard so it now prints out the cell/ports which are responsible for the loop.

This is what is shown now:

```
Info: Found 1 combinational loops:
Info:     loop 1:
Info:         out_LUT4_Z.D (out$TRELLIS_IO_OUT)
Info:         out_LUT4_Z.F (out$TRELLIS_IO_OUT)
Warning: Timing analysis failed due to combinational loops
```